internal/tool/multi/deployer.go

```diff
@@ -198,6 +179,68 @@ func newDeployer(ctx context.Context, deploymentId string, config *protos.AppCon
 	return d, nil
 }
 
+// computeGroups computes the colocation group information for the given
+// application.
+//
+// computeGroups places components into co-location groups based on the
+// colocate stanza in a config. For example, consider the following config.
+//
+//	colocate = [
+//	    ["A", "B", "C"],
+//	    ["D", "E"],
+//	]
+//
+// computeGroups creates a co-location group with components "A", "B", and
+// "C" and a co-location group with components "D" and "E". All other
+// components are placed in their own co-location group. We use the first
+// listed component as the co-location group name.
+func computeGroups(config *protos.AppConfig) (map[string]*group, error) {
+	groups := map[string]*group{}
+	ensureGroup := func(component string) *group {
+		if g, ok := groups[component]; ok {
+			return g
+		}
+		g := &group{
+			name:        component,
+			started:     map[string]bool{},
+			addresses:   map[string]bool{},
+			assignments: map[string]*protos.Assignment{},
+			subscribers: map[string][]*envelope.Envelope{},
+		}
+		groups[component] = g
+		return g
+	}
+
+	// Use the colocation information to place multiple components
+	// in the same group.
+	for _, grp := range config.Colocate {
+		if len(grp.Components) == 0 {
+			continue
+		}
+		g := ensureGroup(grp.Components[0])
+		for i := 1; i < len(grp.Components); i++ {
+			groups[grp.Components[i]] = g
+		}
+	}
+
+	// Use the call graph information to (1) identify all components in the
+	// application binary and create groups for them, and (2) compute the
+	// set of components a given group is allowed to invoke methods on.
+	callGraph, err := bin.ReadComponentGraph(config.Binary)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read the call graph from the application binary: %w", err)
+	}
+	for _, edge := range callGraph {
+		src := edge[0]
+		dst := edge[1]
```

Comment:
Should we change the edge from a [2]string to struct{Src, Dst string}?

Reply:
I like it the way it is: simple.